src/activity_handlers.cpp

```diff
@@ -475,542 +475,183 @@ int butcher_time_to_cut( const player &u, const item &corpse_item, const butcher
-            p->add_msg_if_player( m_good, _( "You harvest some feathers!" ) );
-        }
-    }
-
-    if( wool > 0 ) {
-        if( corpse->has_flag( MF_WOOL ) ) {
-            g->m.spawn_item( p->pos(), "wool_staple", wool, 0, age );
-            p->add_msg_if_player( m_good, _( "You harvest some wool staples!" ) );
-        }
-    }
+void butchery_drops_harvest( item *corpse_item, const mtype &mt, player &p, const time_point &age,
+                             const std::function<int()> &roll_butchery, butcher_type action )
+{
+    p.add_msg_if_player( m_neutral, _( mt.harvest->message().c_str() ) );
+    int monster_weight = to_gram(mt.weight);
 
-    if( fats > 0 ) {
-        if( corpse->has_flag( MF_FAT ) && corpse->has_flag( MF_POISON ) ) {
-            g->m.spawn_item( p->pos(), "fat_tainted", fats, 0, age );
-            p->add_msg_if_player( m_good, _( "You harvest some gooey fat!" ) );
-        } else if( corpse->has_flag( MF_FAT ) ) {
-            g->m.spawn_item( p->pos(), "fat", fats, 0, age );
-            p->add_msg_if_player( m_good, _( "You harvest some fat!" ) );
-        }
+    if (corpse_item->has_flag("QUARTERED")) {
+        monster_weight = monster_weight / 4;
     }
 
-    //Add a chance of CBM recovery.
-    //As long as the factor is above -4 (the sinew cutoff), you will be able to extract CBMs
-    if( action == DISSECT ) {
-
-        if( corpse->has_flag( MF_CBM_CIV ) ) {
-            butcher_cbm_item( "bio_power_storage", p->pos(), age, roll_butchery() );
-            butcher_cbm_group( "bionics_common", p->pos(), age, roll_butchery() );
-        }
-
-        // Zombie scientist bionics
-        if( corpse->has_flag( MF_CBM_SCI ) ) {
-            butcher_cbm_item( "bio_power_storage", p->pos(), age, roll_butchery() );
-            butcher_cbm_group( "bionics_sci", p->pos(), age, roll_butchery() );
-        }
-
-        // Zombie technician bionics
-        if( corpse->has_flag( MF_CBM_TECH ) ) {
-            butcher_cbm_item( "bio_power_storage", p->pos(), age, roll_butchery() );
-            butcher_cbm_group( "bionics_tech", p->pos(), age, roll_butchery() );
-        }
-
-        // Substation mini-boss bionics
-        if( corpse->has_flag( MF_CBM_SUBS ) ) {
-            butcher_cbm_item( "bio_power_storage", p->pos(), age, roll_butchery() );
-            butcher_cbm_group( "bionics_subs", p->pos(), age, roll_butchery() );
-            butcher_cbm_group( "bionics_subs", p->pos(), age, roll_butchery() );
+    int monster_weight_remaining = monster_weight;
+    int practice = 4 + roll_butchery();
+    for (const auto &entry : *mt.harvest) {
+        int butchery = roll_butchery();
+        float min_num = entry.base_num.first + butchery * entry.scale_num.first;
+        float max_num = entry.base_num.second + butchery * entry.scale_num.second;
+        int roll = 0;
+        // mass_ratio is not strictly required, and if it is not there, it allows for more precisely define amounts of items butchered
```

Comment:
If it's not strictly required, it's optional. Please wrap it into `cata::optional` instead of using a meaningless and undocumented sentinel value.

Reply:
i'm afraid i don't know what you mean by wrapping it into `cata::optional` nor do i know what a sentinel value is. the comment there that says it's not strictly required is saying that it's not required to be defined in the JSON entry, if that changes what you mean